### PR TITLE
[CodeGen][Common] Switch to new pass generation tablegen definitions. 

### PR DESCRIPTION
**Pull request by hanhanW**
This is not a NFC change because there are non-trivial changes in three passes. The rest are just switching to the new tablegen definition style.

The revision applies few cleanups during refactoring, which includes:

- Make all the passes have `iree-codegen-*` prefix.
- Make filename match the pass name.
- Switch namespaces to the single-line syntax.

The non-trivial changes happen in `IREEComprehensiveBufferizePass`, `TileAndDistributeToWorkgroupsPass`, and `TransformDialectInterpreterPass`.

In the `IREEComprehensiveBufferizePass`, the default allocationFn and memcpyFn values are no longer `std::nullopt`. They are the default functions that implemented in the file. It matches the behavior. The only difference is that it does not rely on users to pass `std::nullopt` to trigger the "default" behavior. Instead, it initialize the default value of the class members be them. This saves one level of nesting logic.

In the `TileAndDistributeToWorkgroupsPass`, it switches to use `clEnumValN`. Without the change, the values in `.td` are magic numbers. Now we explicitly tie them to linalg::DistributionMethod.

In `TransformDialectInterpreterPass`, it no longer uses `--iree-codegen-transform-dialect-library` to get the library path. Instead, users should use the predefined option (i.e., `library-file-name`) to specify the path. The old behavior was a hack because the pass itself should not access external variable if possible. The `--iree-codegen-transform-dialect-library` is still valid, but the lit test does not use it anymore. Because it is very tricky to teach tablegen to generate such behavior.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/Passes.td [hunk 1]
```diff
@@ -470,70 +453,71 @@ def SplitFullPartialTransfer :
...
+                       "0", "Use Cyclic strategy"),
+            clEnumValN(linalg::DistributionMethod::CyclicNumProcsGeNumIters,
+                       "1", "Use CyclicNumProcGeNumIter strategy"),
+            clEnumValN(linalg::DistributionMethod::CyclicNumProcsEqNumIters,
+                       "2", "Use CyclicNumProcGeNumIter strategy"),
+            clEnumValN(linalg::DistributionMethod::None,
+                       "3", "Use None strategy")
+        )}]>,
   ];
 }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/Passes.td]: This is the non-trivial part for `TileAndDistributeToWorkgroupsPass`. @MaheshRavishankar please take a look.

File: compiler/src/iree/compiler/Codegen/Common/IREEComprehensiveBufferizePass.cpp [hunk 2]
```diff
@@ -88,34 +116,11 @@ class IREEComprehensiveBufferizePass
   void runOnOperation() override;
 
 private:
-  const std::optional<BufferizationOptions::AllocationFn> allocationFn;
-  const std::optional<BufferizationOptions::MemCpyFn> memCpyFn;
+  const BufferizationOptions::AllocationFn allocationFn = defaultAllocationFn;
+  const BufferizationOptions::MemCpyFn memCpyFn = defaultMemCpyFn;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/IREEComprehensiveBufferizePass.cpp:120]: This is the non-trivial part for `IREEComprehensiveBufferizePass` that mentioned in the PR description. @MaheshRavishankar please take a look.

**Review by hanhanW — COMMENTED** [review]
For the transform dialect interpreter pass changes, please take a look at https://github.com/iree-org/iree/pull/18166/commits/ebaafb007ad2e16c0a5cf7ec9cf6b82f42eb896d

@qedawkins please take a look

(cc @Groverkss @raikonenfnu @kuhar because the usage of transform interpreter is changed at `iree-opt` level.)

File: compiler/src/iree/compiler/Codegen/Common/Passes.h [hunk 3]
```diff
@@ -200,104 +74,28 @@ createHoistUnrolledVectorExtractInsertSlicePass();
...
-std::unique_ptr<InterfacePass<FunctionOpInterface>> createTypePropagationPass();
+    int32_t maxWorkgroupParallelDims,
+    linalg::DistributionMethod distributionMethod);
 
-/// Pass to vectorize memref.copy ops.
-std::unique_ptr<OperationPass<void>> createVectorizeMemrefCopyPass();
+// TODO(hanchung): Move it where it is defined (i.e., Codegen/LLVMGPU).
+// Extract address computations (including the ones with GPU instructions) into
+// their own separate instructions.
+std::unique_ptr<Pass> createExtractAddressComputationGPUPass();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/Passes.h:94]: I'll send a follow-up for this after landing the PR.

File: compiler/src/iree/compiler/Codegen/Common/Passes.td (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/Passes.td]: A typo is fixed in https://github.com/iree-org/iree/pull/18166/commits/6db4e408abe2bf76b78257027c7cc342d720cbf6

File: compiler/src/iree/compiler/Codegen/Common/AddFastMathFlags.cpp [hunk 4]
```diff
@@ -30,18 +31,15 @@ namespace {
 /// optimization mode.
 // TODO: For now we only allow default flags, such as arithmetic reassociation.
 struct AddFastMathFlagsPass
-    : public AddFastMathFlagsBase<AddFastMathFlagsPass> {
+    : public impl::AddFastMathFlagsPassBase<AddFastMathFlagsPass> {
 public:
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/AddFastMathFlags.cpp]: nit:
```suggestion
    final : impl::AddFastMathFlagsPassBase<AddFastMathFlagsPass> {
```

File: compiler/src/iree/compiler/Codegen/Common/BubbleUpOrdinalOps.cpp [hunk 5]
```diff
@@ -63,7 +65,7 @@ struct BubbleUpAcrossCastOp
 };
 
 struct BubbleUpOrdinalOpsPass
-    : public BubbleUpOrdinalOpsBase<BubbleUpOrdinalOpsPass> {
+    : public impl::BubbleUpOrdinalOpsPassBase<BubbleUpOrdinalOpsPass> {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/BubbleUpOrdinalOps.cpp]:
```suggestion
    final : impl::BubbleUpOrdinalOpsPassBase<BubbleUpOrdinalOpsPass> {
```

File: compiler/src/iree/compiler/Codegen/Common/BufferizeCopyOnlyDispatchesPass.cpp [hunk 6]
```diff
@@ -30,16 +29,18 @@
...
 
 /// Pass to bufferize early copy-only dispatches. This allows backends
 /// to use the `linalg.generic` operation generated for lowering the dispatch.
 struct BufferizeCopyOnlyDispatchesPass
-    : public BufferizeCopyOnlyDispatchesBase<BufferizeCopyOnlyDispatchesPass> {
-  BufferizeCopyOnlyDispatchesPass() = default;
-  BufferizeCopyOnlyDispatchesPass(const BufferizeCopyOnlyDispatchesPass &pass) {
-  }
-
+    : public impl::BufferizeCopyOnlyDispatchesPassBase<
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/BufferizeCopyOnlyDispatchesPass.cpp]:
```suggestion
    final : impl::BufferizeCopyOnlyDispatchesPassBase<
```

File: compiler/src/iree/compiler/Codegen/Common/CleanupBufferAllocViewPass.cpp [hunk 7]
```diff
@@ -26,11 +25,14 @@
...
 
+#define GEN_PASS_DEF_CLEANUPBUFFERALLOCVIEWPASS
+#include "iree/compiler/Codegen/Common/Passes.h.inc"
+
 namespace {
 
 /// Runs canonicalization patterns on interface load/store ops.
 struct CleanupBufferAllocViewPass
-    : public CleanupBufferAllocViewBase<CleanupBufferAllocViewPass> {
+    : public impl::CleanupBufferAllocViewPassBase<CleanupBufferAllocViewPass> {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/CleanupBufferAllocViewPass.cpp]:
```suggestion
    final : impl::CleanupBufferAllocViewPassBase<CleanupBufferAllocViewPass> {
```

File: compiler/src/iree/compiler/Codegen/Common/ConcretizePadResultShape.cpp [hunk 8]
```diff
@@ -126,12 +128,9 @@ struct ConcretizePadResultShape final : public OpRewritePattern<tensor::PadOp> {
 };
 
 class ConcretizePadResultShapePass final
-    : public ConcretizePadResultShapeBase<ConcretizePadResultShapePass> {
+    : public impl::ConcretizePadResultShapePassBase<
+          ConcretizePadResultShapePass> {
 public:
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/ConcretizePadResultShape.cpp:133]:
```suggestion
   final : impl::ConcretizePadResultShapePassBase<
          ConcretizePadResultShapePass> {
```

File: compiler/src/iree/compiler/Codegen/Common/ConvertBf16ArithToF32.cpp [hunk 9]
```diff
@@ -236,8 +238,9 @@ struct PromoteBF16ToF32Converter
 };
 
 struct ConvertBf16ArithToF32Pass
-    : public ConvertBf16ArithToF32Base<ConvertBf16ArithToF32Pass> {
-  using ConvertBf16ArithToF32Base::ConvertBf16ArithToF32Base;
+    : public impl::ConvertBf16ArithToF32PassBase<ConvertBf16ArithToF32Pass> {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/ConvertBf16ArithToF32.cpp]: Similar here and everywhere else

**Review by kuhar — COMMENTED** [review]
Awesome, thanks for modernizing these!

Only one nit from me.

File: compiler/src/iree/compiler/Codegen/Common/IREEComprehensiveBufferizePass.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Common/IREEComprehensiveBufferizePass.cpp:120]: Actually I am confused what happens here. When is the default picked up ? Also the naming makes things hard to follow. Can you make the names different (from just `allocationFn` everywhere)

**Review by raikonenfnu — COMMENTED** [review]
Not much comments from me aside from thanks for this super awesome clean up/ modernization @hanhanW

File: compiler/src/iree/compiler/Codegen/Common/IREEComprehensiveBufferizePass.cpp [hunk 10]
```diff
@@ -57,13 +83,15 @@ class EliminateEmptyTensorsPass
...
+    : public impl::IREEComprehensiveBufferizePassBase<
+          IREEComprehensiveBufferizePass> {
 public:
+  using impl::IREEComprehensiveBufferizePassBase<
+      IREEComprehensiveBufferizePass>::IREEComprehensiveBufferizePassBase;
   explicit IREEComprehensiveBufferizePass(
-      std::optional<BufferizationOptions::AllocationFn> allocationFn =
-          std::nullopt,
-      std::optional<BufferizationOptions::MemCpyFn> memCpyFn = std::nullopt)
+      BufferizationOptions::AllocationFn allocationFn,
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/IREEComprehensiveBufferizePass.cpp:93]: Can we do 
```
IREEComprehensiveBufferizePass(
    BufferizationOptions::AllocationFn allocationFn = defaultAllocationFn,
    BufferizationOptions::AllocationFn memCpyFn = defaultMemCpyFn) : 
 allocationFn(allocationFn), memCpyFn(memCpyFn) {}
```

**Review by MaheshRavishankar — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/IREEComprehensiveBufferizePass.cpp (hunk 10 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/IREEComprehensiveBufferizePass.cpp:93]: Discussed it offline, it is not feasible because it conflicts with the constructor generated by tablegen.

**Review by kuhar — APPROVED** [review]

**Review by qedawkins — APPROVED** [review]
LGTM, I don't think the transform dialect changes will affect any downstream flows either, IIUC
